### ISSUE TITLE
Setup Command: pass package as command

### ISSUE THREAD

**Issue by thedavidprice** [redwoodjs/redwood#1930]
_Update 3/9/21:_ 
See [this comment below](https://github.com/redwoodjs/redwood/issues/1930#issuecomment-794716342) for decisions and next steps

---

This Issue consolidates the discussion orginally started as a proposal on the Forum:
- [RW Setup Proposal: Complementing Generators with Setup](https://community.redwoodjs.com/t/rw-setup-proposal-complementing-generators-with-setup/1123)

Issue https://github.com/redwoodjs/redwood/issues/1323 included a POC by @dac09 passing a plugin (node package):
- See [this comment](https://github.com/redwoodjs/redwood/issues/1323#issuecomment-707788829)

## Background
Included in Redwood's v1 Roadmap is the Setup command, which is used for one-time actions involving installation, instantiation, boilerplate, and configuration. The Setup command has already been created in part using Yargs built-in Command Directory feature.

However, the initial concept of the Setup command included the capability to pass a set of sub-commands at once, allowing for a project "recipe" that would quickly, in one step, configure a Redwood project. 

The purpose of this Issue is to finalize the architecture of the setup command in order to complete implementation.

## Current Requirements (aggregated from previous discussions)
- allow yarn rw setup [options] to take an array of commands
- accept "global" flags, e.g. `--force`
- accept flags specific to a sub-command
- accept packages as a sub-command (or a sub-command that accepts packages... ?)
  - at one time it was also suggested a path could be used
- namespace: how to structure command names, e.g. [category]- [provider]-[misc] like `auth-netlify` or `deploy-netlify` (note: this might be handled by command syntax)
- `yarn rw setup --help` should output, well, something helpful
  - maybe we create categories with lists
- `yarn rw setup [sub-command] --help` should output help specific to the sub-command

## Next Steps
This Issue is open to discussion + feedback from the community. Ultimately, the Core Team will finalize the requirements and architecture, then select an individual to lead the PR effort in a collaborative fashion.

**Comment by thedavidprice** [comment]
_copied from original comment by @jtoar from here https://github.com/redwoodjs/redwood/pull/1408#issuecomment-782750006

---

To show what we're up against, here's a step-by-step description of tweaking `setup` to better accept @olance's command string. I made it so that `setup` just returns `argv`. So setup looks like this:

```js
export const command = 'setup <commmand>'
export const description = 'Initialize project config and install packages'
export const handler = (argv) => console.log(argv)
```

`argv` is what the yargs-parser gives us from what the user enters at the command line. Let's see what yargs does by default with @olance's command string  

```
yarn rw setup --force +deploy:netlify +auth:supabase +tailwind --ui +i18n --fr
```

```js
{
  _: [ 'setup', '+tailwind' ],
  force: '+deploy:netlify',
  ui: '+i18n',
  fr: true,
  '$0': 'rw',
  commmand: '+auth:supabase'
}
```

There's a lot of problems with the way this is being parsed. For one, `--force` is receiving `+deploy:netlify` as it's argument (instead of it just being a boolean, like `force: true`)&mdash;we can quickly fix that by adding a builder to `setup` that specifies the options it takes:

```diff
 export const command = 'setup <commmand>'
 export const description = 'Initialize project config and install packages'
+export const builder = (yargs) => yargs.option('force', { type: 'boolean' })
 export const handler = (argv) => console.log(argv)
```

```js
{
  _: [ 'setup', '+auth:supabase', '+tailwind' ],
  force: true,
  ui: '+i18n',
  fr: true,
  '$0': 'rw',
  commmand: '+deploy:netlify'
}
```

Better, but `+deploy:netlify`'s in `command` while everything else is in `_` (only `setup` should really be there). Let's try changing `setup`'s `command` to accept variadic positional arguments:

```diff
-export const command = 'setup <commmand>'
+export const command = 'setup [commmand..]'
 export const description = 'Initialize project config and install packages'
 export const builder = (yargs) => yargs.option('force', { type: 'boolean' })
 export const handler = (argv) => console.log(argv)
```

```js
{
  _: [ 'setup' ],
  force: true,
  ui: '+i18n',
  fr: true,
  '$0': 'rw',
  commmand: [ '+deploy:netlify', '+auth:supabase', '+tailwind' ]
}
```

Much better! But we still haven't addressed that `ui`, an option to `+tailwind`, is receiving `+i18n` as it's argument; `+i18n` should really be in `command`, and `ui` should be `true` like `force`.

This is where we start to run up against the limit of what we can do just by tweaking the yargs exports (command/description/builder/handler). So where do we go from here?

### make it so that the builder knows all the subcommand's options

Why don't we just add `ui` to `builder` like we did with `force`? That would make it so that it parses as `true`. But now we have to ask: how do we know `--ui` is for `+tailwind`? There's nothing in `argv` that says so; we'd just have to know in advance. And the problem there is we can't have subcommands that have the same options. This is a huge problem with `--force`, which pretty much everything accepts. How do we know what you wanted to overwrite and what you didn't? 

### write some gnarly middleware to reparse argv before it hits the handler

Yargs has the concept of [middleware](https://github.com/yargs/yargs/blob/master/docs/advanced.md#middleware); it runs before `argv` hits the handler, so we could do something kind of similar to `afterQuery` in a Cell, and majorly restructure `argv`. But that still doesn't address the option clashing issue.

### organizing arguments into nested objects

[Quoting from the Yargs docs](https://github.com/yargs/yargs/blob/master/docs/tricks.md#objects),

> When you use dots (.s) in argument names, an implicit object path is assumed. This lets you organize arguments into nested objects.

To see this in action, let's tweak @olance's command string and run it:

```diff
-yarn rw setup --force +deploy:netlify +auth:supabase +tailwind --ui +i18n --fr
+yarn rw setup --force --deploy netlify --auth supabase --tailwind.ui --i18n.fr
```

```js
  _: [ 'setup' ],
  force: true,
  deploy: 'netlify',
  auth: 'supabase',
  tailwind: { ui: true },
  i18n: { fr: true },
  '$0': 'rw'
}
```

That actually looks amazing! We can just iterate through the object properties; the subcommands are already organized with their arguments.

To really drive the point home, let's add an option to `deploy`, like `--data-migrate`:

```diff
-yarn rw setup --force --deploy netlify --auth supabase --tailwind.ui --i18n.fr
+yarn rw setup --force --deploy netlify --deploy.data-migrate --auth supabase --tailwind.ui --i18n.fr
```

```js
{
  _: [ 'setup' ],
  force: true,
  deploy: [ 'netlify', { 'data-migrate': true }, { dataMigrate: true } ],
  auth: 'supabase',
  tailwind: { ui: true },
  i18n: { fr: true },
  '$0': 'rw'
}
```

That looks near perfect. And our setup command looks like this; it's as vanilla as it gets (I even took out `builder` and `force` _still_ parses correctly)!

```js
export const command = 'setup'
export const description = 'Initialize project config and install packages'
export const handler = (argv) => console.log(argv)
```

> Did we have to repeat `--deploy`? Yes, and that might be a shortcoming of this syntax.

### conclusion

So I'm not sure what to do from here; I just wanted to illustrate that there's a "native" syntax yargs understands. Whether or not we like that syntax is the question now.

As @mojombo says, computers should do what we want, so we should probably just pick the syntax we like and code for it. But with the dot syntax, it's a downhill battle/most of the work's already done.

**Comment by thedavidprice** [comment]
_copied from original comment by @olance here https://github.com/redwoodjs/redwood/pull/1408#issuecomment-786741110_

---

I've setup a PoC repo here: https://github.com/olance/poc-redwood-yargs-parsing

With the following params:

```
--force +deploy:netlify --migrate-data +auth:supabase +tailwind --ui --version 2 +i18n --lang=en,fr
```

The `argv` object out of `yargs` is what you obtained using the dotted notation:

```js
{
  force: true,
  deploy: [ 'netlify', { 'migrate-data': true }, { migrateData: true } ],
  auth: 'supabase',
  tailwind: { ui: true, version: 2 },
  i18n: { lang: 'en,fr' },
}
```

The code is pretty simple, it only walks through the args and transforms them to the dotted notation `yargs` supports _before_ it is parsed:

```js
const argv = yargs(processArgs(hideBin(process.argv))).argv
```

```js
function processArgs(argv) {
  console.log("Input argv:", argv)

  let args = argv.reverse()
  let currentScope = null
  let newArgs = []

  while (args.length > 0) {
    let arg = args.pop()

    // Beginning a new scope?
    if (arg[0] === SCOPE_TOKEN) {
      // Extract scope name.
      // That would be 'deploy' for '+deploy:netlify' and 'tailwind' for '+tailwind'
      // TODO: should we handle multi-level scopes, i.e. +deploy:gcp:cloud-run / +deploy:gcp:appengine? (seems unlikely)
      const scopeComponents = arg.substr(1).split(SCOPE_VALUE_SEPARATOR)
      currentScope = scopeComponents[0]

      // If it's a "parameterized" scope ('+deploy:netlify') we need to issue an arg for the scope value
      if (scopeComponents.length === 2) {
        newArgs.push(`--${currentScope}=${scopeComponents[1]}`)
      }

      continue
    }

    // Down here means we're _within_ a scope, possibly null (global scope)
    const argName = arg.substr(2) // Assuming args always start with '--'
    let newArg = '--'

    if (currentScope !== null) {
      newArg += `${currentScope}.`
    }

    // We now have newArg ==
    // -> '--force' for a global '--force' flag
    // -> '--tailwind.ui' for something like '+tailwind --ui'
    // -> '--i18n.lang=en,fr' for something like '+i18n --lang=en,fr'
    newArg += argName

    // If the arg needs a value and it's been passed like '--lang en,fr' it's not covered by the above, we need to look
    // ahead and consume the next arg if needed
    const nextArg = args[args.length - 1]
    if (argName.indexOf('=') < 0 && nextArg[0] !== SCOPE_TOKEN && nextArg.substr(0, 2) !== '--') {
      newArg += `=${args.pop()}`
    }

    newArgs.push(newArg)
  }

  console.log('Transformed argv:', newArgs)

  return newArgs
}
```

**Comment by Tobbe** [comment]
I like the "+" syntax

And it should still support shorthand options, like -f for --force, right?

So you could do this
```
yarn rw setup +deploy:netlify -f --data-migrate +auth:supabase +tailwind --ui +i18n --fr
```
And it would only add `--force` to the deploy command (together with `--data-migrate`)

Why do we need the ":" syntax again? Why can't it just be `+auth --supabase`? Or, how do you know it's `+tailwind --ui` and not `+tailwind:ui`. I just think that can be a source of confusion

**Comment by dac09** [comment]
I've already spoken to Dom about this one, but I really think passing flags like that to the subcommands is wrong. For once I feel (a little) strongly about this. I love CLI tools and complexity doesn't scare me, _as long as_ the command/flag behaviour is consistent.

IMHO flags should only apply to the first command i.e. in this case setup, and this avoids a huge amount of confusion. This is consistent and expected behaviour. FFMpeg is a great example of a tool where you can pass 500 flags, but the anything that is flag i.e. `--flag 1` or `-c 26` always applies to the first command.

Let's take this command for example (separated into new lines for readability):
```
ffmpeg \
-i ipgrey.png \
-i input.mp4 \
-filter_complex "\
[1:v]scale=1135:-1,pad=w=100+iw:200+ih:(ow-iw)/2:(oh-ih)/2:color=white[scaled_video],\
[scaled_video][0:v]\
overlay=x=(main_w-overlay_w)/3:y=(main_h-overlay_h)/3" \
-preset faster \
-y wrapped.mp4
```
For me in this relatively complicated command, when I'm adding even more flags, its easy to reason about what the flag applies to - its always applying to ffmpeg, not to_filter complex. If I wanted to pass a flag to a "sub command", like filter-complex, I'd have to include it in the filter string.

**Tldr;** 
the syntax should perhaps look like this if we're passing flags:
```
yarn rw setup "deploy netlify -f --data-migrate" "auth supabase" "tailwind --ui" 
```
And as a convenience we could potentially do this when no flags are required
```
yarn rw setup deploy:netlify auth:supabase tailwind
```

**Comment by Tobbe** [comment]
That works too. Still a little worried it's difficult to know if it's `--` or not in front of args.
`auth supabase` vs `auth --supabase`
`tailwind ui` vs `tailwind --ui`
But maybe we can't do anything about that, and one just has to read the docs to know what's positional args and what's options.

**Comment by mojombo** [comment]
These proposed syntaxes are way too complicated. I think it only makes sense to allow multiple setups at once if they do not take arguments (like `yarn add`). Once our setup commands need to take arguments, we should drop the ability to do multiple in a single invocation and just require multiple invocations. That would fix all these problems, yes?

**Comment by peterp** [comment]
Is combining setup scripts into a single execution something that we should be doing?

I don't think so because the pre-install requirements and post-install completion steps for setup commands are very different, and finishing a setup command often requires comprehension and intervention from the user.

As an example, If I wanted to setup `deploy aws-serverless` and `auth auth0`
![image](https://user-images.githubusercontent.com/44849/110139359-9cfbf580-7ddb-11eb-915c-bd8d89b95fd3.png)
![image](https://user-images.githubusercontent.com/44849/110139286-8d7cac80-7ddb-11eb-9f5a-f994f2c7eb7d.png)

The first command failed because I need to install something, and the second command requires that I read documentation, create an account, and create env-vars.

My problem is that I don't know how we could effectively communicate all of this to the user. In the first case they would have to fix the requirements and re-run the command, but in the second case they've already succeeded in running the command.

Of course, we could terminate the script if we detect a single error, but what if it breaks in the middle of a chain? Our setup commands would have to be self-aware in order to understand if they've run or not (which is also probably ok to do), like re-running the `auth auth0` command spits out an error:

![image](https://user-images.githubusercontent.com/44849/110140496-d3864000-7ddc-11eb-9c69-8f134ee5d569.png)

Which we could turn that into a warning, but then the user needs to comprehend all of that to understand what worked, what didn't, and what requires intervention.

I just don't think there are enough gains here for the user.

**Comment by thedavidprice** [comment]
Per Core Team discussion, we are moving forward with the simple path:
1. keep the existing structure, which allows for one command with options, e.g. `...setup tailwind --force`
2. add the capability to provide a compatible package as the command (see example by @dac09)

## Next Steps
- [ ] RFC for accepting a package
- [ ] Discussion: pass a URL as a command? @jtoar

**Comment by Tobbe** [comment]
Purely a nice-to-have, and absolutely not a must-have. But it would be cool to be able to pass in a url or package with custom templates. Like a custom template for `yarn rw g page` to use, and another for `yarn rw g scaffold` etc. Then, as a company, I can provide templates to make sure all apps I spin up share the same basic look-and-feel.

**Comment by thedavidprice** [comment]
@dac09 reminder to close this Issue out and replace it with new, consolidated RFC based on your POC. Thanks!

**Comment by dac09** [comment]
Closing this in favour of https://github.com/redwoodjs/redwood/issues/2086

Key decision: we won't be supporting multiple setup commands in a single command right now
